engine-core/src/main/java/io/nosqlbench/engine/core/lifecycle/activity/ActivityExecutor.java

```diff
@@ -22,43 +22,48 @@
 import io.nosqlbench.engine.core.annotation.Annotators;
+import io.nosqlbench.engine.core.lifecycle.ExecutionResult;
+import io.nosqlbench.engine.core.lifecycle.IndexedThreadFactory;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-import java.util.*;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.SynchronousQueue;
-import java.util.concurrent.ThreadPoolExecutor;
-import java.util.concurrent.TimeUnit;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.*;
 import java.util.stream.Collectors;
 
 /**
- * <p>An ActivityExecutor is a named instance of an execution harness for a single activity instance.
+ * <p>An ActivityExecutor is an execution harness for a single activity instance.
  * It is responsible for managing threads and activity settings which may be changed while the activity is running.</p>
  *
- * <p>An ActivityExecutor may be represent an activity that is defined and active in the running
- * scenario, but which is inactive. This can occur when an activity is paused by controlling logic, or when the threads
- * are set to zero.</p>
+ * <p>In order to allow for dynamic thread management, which is not easily supported as an explicit feature
+ * of most executor services, threads are started as long-running processes and managed via state signaling.
+ * The {@link RunState} enum, {@link io.nosqlbench.engine.api.activityimpl.MotorState} type, and {@link RunStateTally}
+ * state tracking class are used together to represent valid states and transitions, contain and transition state atomically,
+ * and provide blocking conditions for observers, respectively.</p>
+ *
+ * <P>Some basic rules and invariants must be observed for consistent concurrent behavior.
+ * Any state changes for a Motor must be made through {@link Motor#getState()}.
+ * This allows the state tracking to work consistently for all observers.</p>
  *
- * <p>
- * Invariants:
- * </p>
- * <ul>
- *     <li>Motors may not receive parameter updates before their owning activities are initialized.</li>
- * </ul>
  */
 
-public class ActivityExecutor implements ActivityController, ParameterMap.Listener, ProgressCapable {
+public class ActivityExecutor implements ActivityController, ParameterMap.Listener, ProgressCapable, Callable<ExecutionResult> {
+
+    // TODO Encapsulate valid state transitions to be only modifiable within the appropriate type view.
 
     private static final Logger logger = LogManager.getLogger(ActivityExecutor.class);
     private static final Logger activitylogger = LogManager.getLogger("ACTIVITY");
 
     private final List<Motor<?>> motors = new ArrayList<>();
     private final Activity activity;
     private final ActivityDef activityDef;
-    private final ExecutorService executorService;
-    private RuntimeException stoppingException;
+    private final RunStateTally tally;
+    private ExecutorService executorService;
+    private Exception exception;
```

Comment:
This is one of the things which was improved, but not enough to call it done. The exception is what is set as a status indicator for success or failure of the related Activity. While it may be possible to set it via multiple mechanisms (multiple sources of exceptions to be thrown, and with different levels of severity), there is no clarity yet on how we record and present these. The most important distinction here is whether there was *some* error, and that should still be tracked reliably, even if one error aliases another.

One of the key changes in this branch was that no matter how the exception is caused, or which one was the "last one" to overwrite this status field, there is only one place from which the official exit status of the activity can be seen, and that is in the future result form the call method. I *do* think we should put something on the plan to review this code and establish much clearer semantics and mechanisms for when multiple errors occur and how to track their cause. There are some corner cases where a less impactful error could occlude a users view of the operative error and that is enough to say this isn't done yet overall.

Reply:
Added #879 to track further changes to this.